tests/subsys/pm/power_domain_soc_state_change/src/main.c

```diff
@@ -0,0 +1,130 @@
+/*
+ * Copyright 2025 NXP
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/ztest.h>
+#include <zephyr/pm/device.h>
+
+#define TEST_DEV DT_NODELABEL(test_dev_soc_state_change)
+
+static int testing_domain_on_notitication;
+static int testing_domain_off_notitication;
+static volatile bool idle_timer_done;
+static struct k_timer idle_timer;
+
+static void idle_timer_expiry_function(struct k_timer *timer_id)
+{
+	k_timer_stop(&idle_timer);
+	idle_timer_done = true;
+}
+
+void pm_state_set(enum pm_state state, uint8_t substate_id)
+{
+	ARG_UNUSED(substate_id);
+
+	idle_timer_done = false;
+
+	/* Start a timer so we return to idle incase we are spuriouly
+	 * woken up.
+	 */
+	k_timer_start(&idle_timer, Z_TIMEOUT_TICKS(2), K_NO_WAIT);
+
+	switch (testing_domain_off_notitication) {
+	case 1:
+		zassert_equal(state, PM_STATE_STANDBY, "Wrong system state %d", state);
+		break;
+	case 2:
+		zassert_true(
+			((state == PM_STATE_SUSPEND_TO_IDLE) || (state == PM_STATE_RUNTIME_IDLE)),
+			"Wrong system state %d", state);
+		break;
+	default:
+		break;
+	}
+
+	do {
+		k_cpu_idle();
+	} while (idle_timer_done == false);
+}
+
+void pm_state_exit_post_ops(enum pm_state state, uint8_t substate_id)
+{
+	ARG_UNUSED(state);
+	ARG_UNUSED(substate_id);
+
+	irq_unlock(0);
+}
+
+
+/* A custom policy manager is used so we can control the Power States
+ * to transition to and test the issuance of TURN_ON and TURN_OFF actions
+ * by the power domain driver.
+ */
+const struct pm_state_info *pm_policy_next_state(uint8_t cpu, int32_t ticks)
+{
+	const struct pm_state_info *cpu_states;
```

Comment:
You should change the policy to use `ticks` to return the appropriate  state or having a better control over the states you are returning. Using `testing_domain_on_notitication` here is fragile since the policy can be called more than once during the sleep due external interruption or another scheduled event.

Reply:
Fixed